File: android/ilds-design-system/src/main/kotlin/com/icicilombard/ilds/components/IldsTabs.kt
package com.icicilombard.ilds.components

import androidx.compose.foundation.BorderStroke
import androidx.compose.foundation.interaction.MutableInteractionSource
import androidx.compose.foundation.interaction.collectIsPressedAsState
import androidx.compose.foundation.layout.Arrangement
import androidx.compose.foundation.layout.Column
import androidx.compose.foundation.layout.Row
import androidx.compose.foundation.layout.defaultMinSize
import androidx.compose.foundation.layout.fillMaxWidth
import androidx.compose.foundation.layout.padding
import androidx.compose.foundation.shape.RoundedCornerShape
import androidx.compose.material3.HorizontalDivider
import androidx.compose.material3.Surface
import androidx.compose.material3.Text
import androidx.compose.runtime.Composable
import androidx.compose.runtime.Immutable
import androidx.compose.runtime.getValue
import androidx.compose.runtime.mutableIntStateOf
import androidx.compose.runtime.remember
import androidx.compose.runtime.setValue
import androidx.compose.ui.Alignment
import androidx.compose.ui.Modifier
import androidx.compose.ui.graphics.Color
import androidx.compose.ui.semantics.Role
import androidx.compose.ui.semantics.role
import androidx.compose.ui.semantics.selected
import androidx.compose.ui.semantics.semantics
import androidx.compose.ui.text.TextStyle
import androidx.compose.ui.unit.Dp
import androidx.compose.ui.unit.dp
import androidx.compose.ui.unit.sp
import com.icicilombard.ilds.tokens.IldsTokens

// Figma set 17667:2334 — mirrors web/Tabs.tsx + lib/ilds_tab.dart.

enum class IldsTabEmphasis { High, Medium }
enum class IldsTabAlignment { Left, Center }

data class IldsTabItem(
    val label: String,
    val icon: (@Composable () -> Unit)? = null,
    val isDisabled: Boolean = false,
)

@Composable
fun IldsTabs(
    tabs: List<IldsTabItem>,
    modifier: Modifier = Modifier,
    selectedIndex: Int? = null,
    defaultSelectedIndex: Int = 0,
    emphasis: IldsTabEmphasis = IldsTabEmphasis.High,
    alignment: IldsTabAlignment = IldsTabAlignment.Left,
    onChange: (Int) -> Unit = {},
) {
    var internalIndex by remember { mutableIntStateOf(defaultSelectedIndex) }
    val activeIndex = selectedIndex ?: internalIndex
    val isHigh = emphasis == IldsTabEmphasis.High

    fun select(index: Int) {
        if (tabs[index].isDisabled) return
        if (selectedIndex == null) internalIndex = index
        onChange(index)
    }

    Column(modifier = modifier) {
        Row(
            modifier = Modifier.fillMaxWidth(),
            horizontalArrangement = when {
                isHigh && alignment == IldsTabAlignment.Center -> Arrangement.Center
                isHigh -> Arrangement.spacedBy(IldsTokens.sp8)
                alignment == IldsTabAlignment.Center -> Arrangement.Center
                else -> Arrangement.Start
            },
        ) {
            tabs.forEachIndexed { index, tab ->
                IldsTabCell(
                    tab = tab,
                    selected = index == activeIndex,
                    emphasis = emphasis,
                    onClick = { select(index) },
                    modifier = if (!isHigh) Modifier.weight(1f) else Modifier,
                )
            }
        }
        if (!isHigh) {
            HorizontalDivider(color = IldsTokens.neutralCoolgray200, thickness = 1.dp)
        }
    }
}

@Composable
private fun IldsTabCell(
    tab: IldsTabItem,
    selected: Boolean,
    emphasis: IldsTabEmphasis,
    onClick: () -> Unit,
    modifier: Modifier = Modifier,
) {
    val interactionSource = remember { MutableInteractionSource() }
    val isPressed by interactionSource.collectIsPressedAsState()
    val colors = remember(selected, tab.isDisabled, isPressed, emphasis) {
        IldsTabColors.resolve(selected, tab.isDisabled, isPressed, emphasis)
    }

    if (emphasis == IldsTabEmphasis.High) {
        Surface(
            onClick = onClick,
            modifier = modifier.semantics {
                role = Role.Tab
                this.selected = selected
            },
            enabled = !tab.isDisabled,
            interactionSource = interactionSource,
            shape = RoundedCornerShape(IldsTokens.radiusLarge),
            color = colors.background,
            border = BorderStroke(1.dp, colors.border),
        ) {
            TabContent(tab, selected, colors.text, emphasis)
        }
    } else {
        Surface(
            onClick = onClick,
            modifier = modifier.semantics {
                role = Role.Tab
                this.selected = selected
            },
            enabled = !tab.isDisabled,
            interactionSource = interactionSource,
            color = Color.Transparent,
            border = if (selected) {
                BorderStroke(3.dp, colors.border)
            } else {
                BorderStroke(0.dp, Color.Transparent)
            },
            shape = RoundedCornerShape(0.dp),
        ) {
            TabContent(tab, selected, colors.text, emphasis)
        }
    }
}

@Composable
private fun TabContent(
    tab: IldsTabItem,
    selected: Boolean,
    textColor: Color,
    emphasis: IldsTabEmphasis,
) {
    Row(
        modifier = Modifier
            .defaultMinSize(minHeight = 36.dp)
            .padding(horizontal = if (emphasis == IldsTabEmphasis.High) IldsTokens.sp32 else IldsTokens.sp12),
        horizontalArrangement = Arrangement.spacedBy(IldsTokens.sp8),
        verticalAlignment = Alignment.CenterVertically,
    ) {
        tab.icon?.invoke()
        Text(
            text = tab.label,
            style = TextStyle(
                fontSize = IldsTokens.fontSize14,
                lineHeight = 16.sp,
                fontWeight = if (selected) IldsTokens.fontWeightBold else IldsTokens.fontWeightMedium,
            ),
            color = textColor,
        )
    }
}

@Immutable
private data class IldsTabColors(
    val background: Color,
    val border: Color,
    val text: Color,
) {
    companion object {
        fun resolve(
            selected: Boolean,
            isDisabled: Boolean,
            isPressed: Boolean,
            emphasis: IldsTabEmphasis,
        ): IldsTabColors {
            if (isDisabled) {
                return IldsTabColors(
                    IldsTokens.globalWhite000,
                    IldsTokens.neutralCoolgray200,
                    IldsTokens.neutralCoolgray300,
                )
            }
            return if (emphasis == IldsTabEmphasis.High) {
                if (selected) {
                    IldsTabColors(
                        IldsTokens.primaryOrange500,
                        IldsTokens.primaryOrange500,
                        IldsTokens.globalWhite000,
                    )
                } else {
                    IldsTabColors(
                        if (isPressed) IldsTokens.neutralCoolgray50 else IldsTokens.globalWhite000,
                        IldsTokens.neutralCoolgray200,
                        if (isPressed) IldsTokens.neutralCoolgray900 else IldsTokens.neutralCoolgray800,
                    )
                }
            } else {
                if (selected) {
                    IldsTabColors(
                        Color.Transparent,
                        IldsTokens.primaryOrange500,
                        IldsTokens.primaryOrange500,
                    )
                } else {
                    IldsTabColors(
                        Color.Transparent,
                        Color.Transparent,
                        if (isPressed) IldsTokens.neutralCoolgray900 else IldsTokens.neutralCoolgray800,
                    )
                }
            }
        }
    }
}

File: ios/Sources/ILDSDesignSystem/IldsTabs.swift
import ILDSTokens
import SwiftUI

// Mirrors lib/ilds_tab.dart + web Tabs.tsx. TODO(web/desktop): hover states — mobile-first defer.

public enum IldsTabEmphasis: Sendable {
    case high, medium
}

public enum IldsTabAlignment: Sendable {
    case left, center
}

public struct IldsTabItem: Identifiable {
    public let id: String
    public let label: String
    public let icon: AnyView?
    public let isDisabled: Bool

    public init(
        id: String = UUID().uuidString,
        label: String,
        icon: (any View)? = nil,
        isDisabled: Bool = false
    ) {
        self.id = id
        self.label = label
        self.icon = icon.map { AnyView($0) }
        self.isDisabled = isDisabled
    }
}

public struct IldsTabs: View {
    private let tabs: [IldsTabItem]
    @Binding private var selectedIndex: Int
    private let emphasis: IldsTabEmphasis
    private let alignment: IldsTabAlignment

    public init(
        tabs: [IldsTabItem],
        selectedIndex: Binding<Int>,
        emphasis: IldsTabEmphasis = .high,
        alignment: IldsTabAlignment = .left
    ) {
        self.tabs = tabs
        _selectedIndex = selectedIndex
        self.emphasis = emphasis
        self.alignment = alignment
    }

    public var body: some View {
        VStack(spacing: 0) {
            ScrollView(.horizontal, showsIndicators: false) {
                HStack(spacing: 0) {
                    if alignment == .center { Spacer(minLength: 0) }
                    ForEach(Array(tabs.enumerated()), id: \.element.id) { index, tab in
                        IldsTabCell(
                            tab: tab,
                            isSelected: selectedIndex == index,
                            emphasis: emphasis,
                            action: {
                                guard !tab.isDisabled else { return }
                                selectedIndex = index
                            }
                        )
                    }
                    if alignment == .center { Spacer(minLength: 0) }
                }
                .overlay(alignment: .bottomLeading) {
                    IldsTabIndicator(
                        tabs: tabs,
                        selectedIndex: selectedIndex,
                        emphasis: emphasis
                    )
                }
            }
            Divider()
                .background(ILDSTokens.neutralCoolgray200)
        }
    }
}

private struct IldsTabCell: View {
    let tab: IldsTabItem
    let isSelected: Bool
    let emphasis: IldsTabEmphasis
    let action: () -> Void

    var body: some View {
        let colors = IldsTabColors.resolve(isSelected: isSelected, isDisabled: tab.isDisabled, emphasis: emphasis)

        Button(action: action) {
            HStack(spacing: ILDSTokens.sp4) {
                if let icon = tab.icon {
                    icon
                        .foregroundStyle(colors.text)
                        .frame(width: ILDSTokens.fontSize14, height: ILDSTokens.fontSize14)
                }
                Text(tab.label)
                    .font(.system(
                        size: ILDSTokens.fontSize14,
                        weight: isSelected ? ILDSTokens.fontWeightBold : ILDSTokens.fontWeightMedium
                    ))
                    .foregroundStyle(colors.text)
            }
            .frame(height: ILDSTokens.sp48)
            .padding(.horizontal, ILDSTokens.sp12)
        }
        .buttonStyle(.plain)
        .disabled(tab.isDisabled)
        .accessibilityLabel(tab.label)
        .accessibilityAddTraits(isSelected ? .isSelected : [])
    }
}

private struct IldsTabIndicator: View {
    let tabs: [IldsTabItem]
    let selectedIndex: Int
    let emphasis: IldsTabEmphasis

    var body: some View {
        GeometryReader { geo in
            let tabWidth = tabs.isEmpty ? 0 : geo.size.width / CGFloat(tabs.count)
            let color = emphasis == .high ? ILDSTokens.primaryOrange500 : ILDSTokens.neutralCoolgray900
            Rectangle()
                .fill(color)
                .frame(width: tabWidth, height: 3)
                .offset(x: tabWidth * CGFloat(selectedIndex))
                .animation(.easeOut(duration: 0.15), value: selectedIndex)
        }
        .frame(height: 3)
    }
}

private struct IldsTabColors {
    let text: Color

    static func resolve(isSelected: Bool, isDisabled: Bool, emphasis: IldsTabEmphasis) -> IldsTabColors {
        if isDisabled {
            return IldsTabColors(text: ILDSTokens.neutralCoolgray300)
        }
        if isSelected {
            let color = emphasis == .high ? ILDSTokens.primaryOrange500 : ILDSTokens.neutralCoolgray900
            return IldsTabColors(text: color)
        }
        return IldsTabColors(text: ILDSTokens.neutralCoolgray400)
    }
}
